File: Estudos/TADs/Convencionais/Lista/Java/Lista.java
package Estudos.TADs.Convencionais.Lista.Java;

public class Lista {
   // Array que representa a lista de elementos
  private int[] lista;
  // Número atual de elementos na lista
  private int n;

  // Construtor padrão que define um tamanho padrão (por exemplo, 6)
  public Lista() {
    this(6);
  }

  // Construtor da Classe que inicializa o array com o tamanho especificado e define `n` como 0
  public Lista(int tamanho) {
    lista = new int[tamanho];
    n = 0;
  }

  // Insere um elemento no início da lista movendo os demais para a direita
  public void inserirInicio(int x) throws Exception {
    if (n >= lista.length) {
      throw new Exception("Erro ao inserir, a lista já está cheia!");
    }

    // Move todos os elementos para a direita, liberando a posição 0
    for (int i = n; i > 0; i--) {
      lista[i] = lista[i - 1];
    }

    // Insere o elemento `x` na primeira posição e incrementa `n`
    lista[0] = x;
    n++;
  }
  
  // Insere um elemento no final da lista
  public void inserirFim(int x) throws Exception {
    if (n >= lista.length) {
      throw new Exception("Erro ao inserir, a lista já está cheia!");
    }
    
    // Adiciona o elemento `x` na posição `n` e incrementa `n`
    lista[n] = x;
    n++;
  }
  
  // Insere um elemento em uma posição específica da lista, movendo elementos para a direita
  // Só é possível se a lista já tiver tido algum elemento inserido (por qualquer um dos 2 métodos anteriores)
  public void inserir(int x, int pos) throws Exception {
    if (n >= lista.length || pos < 0 || pos > n) {
      throw new Exception("Erro ao inserir, posição inválida ou lista cheia!");
    }

    // Move os elementos para a direita para abrir espaço na posição `pos`
    for (int i = n; i > pos; i--) {
        lista[i] = lista[i - 1];
    }

    // Insere o elemento `x` na posição especificada e incrementa `n`
    lista[pos] = x;
    n++;
  }

  // Remove o elemento da primeira posição, deslocando os outros para a esquerda
  public int removerInicio() throws Exception {
    if (n == 0) {
      throw new Exception("Erro ao remover, a lista está vazia!");
    }

    // Armazena o elemento a ser removido
    int resp = lista[0];
    n--;

    // Move todos os elementos para a esquerda
    for (int i = 0; i < n; i++) {
      lista[i] = lista[i + 1];
    }

    return resp;
  }

  // Remove o último elemento da lista, reduzindo `n`
  public int removerFim() throws Exception {
    if (n == 0) {
      throw new Exception("Erro ao remover, a lista está vazia!");
    }

    // Decrementa `n` e retorna o elemento removido
    return lista[--n];
  }

   // Remove o elemento na posição especificada e move os elementos restantes para a esquerda
  public int remover(int pos) throws Exception {
    if (n == 0 || pos < 0 || pos >= n) {
      throw new Exception("Erro ao remover, posição inválida ou lista vazia!");
    }

    // Armazena o elemento a ser removido
    int resp = lista[pos];
    n--;

    // Move os elementos para a esquerda a partir da posição `pos`
    for (int i = pos; i < n; i++) {
      lista[i] = lista[i + 1];
    }

    return resp;
  }

  // Exibe todos os elementos da lista entre colchetes
  public void mostrar(){
    System.out.print("[ ");
    for (int i = 0; i < n; i++) {
      System.out.print(lista[i] + " ");
    }
    System.out.println("]");
  }

  // Verifica se o elemento `x` está presente na lista, retornando true ou false
  public boolean pesquisar(int x) {
    for (int i = 0; i < n; i++) {
      if (lista[i] == x) {
        return true;
      }
    }
    return false;
  }
}

File: Estudos/TADs/Convencionais/Lista/C/lista.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#define MAX_TAM 10 // Tamanho máximo da lista

int lista[MAX_TAM]; // Array para armazenar os números
int n = 0; // Contador de elementos na lista

/**
 * Insere um número na primeira posição da lista e move os demais elementos para o fim.
 * @param elemento Valor do número a ser inserido.
 */
void inserirInicio(int elemento) {

   // Verifica se há espaço para inserir um novo número
   if (n >= MAX_TAM) {
      printf("Erro ao inserir! Lista cheia.\n"); // Mensagem de erro se a lista está cheia
      exit(1); // Encerra o programa em caso de erro
   } 

    // Move os elementos existentes para a direita, abrindo espaço no índice 0
    for (int i = n; i > 0; i--) {
      lista[i] = lista[i-1]; // Desloca cada número uma posição para a direita
    }

   // Insere o novo número na primeira posição
   lista[0] = elemento;
   n++; // Incrementa o contador de elementos  
}

/**
 * Insere um número na última posição da lista.
 * @param elemento Valor do número a ser inserido.
 */
void inserirFim(int elemento) {

   // Verifica se há espaço para inserir um novo número
    if (n >= MAX_TAM) {
        printf("Erro ao inserir! Lista cheia.\n"); // Mensagem de erro se a lista está cheia
        exit(1); // Encerra o programa em caso de erro
    }

    // Insere o novo número na última posição disponível
    lista[n] = elemento; 
    n++; // Incrementa o contador de elementos
}

/**
 * Insere um número em uma posição específica e move os demais elementos para o fim.
 * @param elemento Valor do número a ser inserido.
 * @param pos Posição onde o número será inserido.
 */
void inserir(int elemento, int pos) {

   // Verifica se a inserção é possível
    if (n >= MAX_TAM || pos < 0 || pos > n) {
        printf("Erro ao inserir! Posição inválida ou lista cheia.\n"); // Mensagem de erro para lista cheia ou posição inválida
        exit(1); // Encerra o programa em caso de erro
    }

    // Move os elementos existentes para a direita, abrindo espaço na posição `pos`
    for (int i = n; i > pos; i--) {
        lista[i] = lista[i - 1]; // Desloca cada número uma posição para a direita
    }

    // Insere o novo número na posição especificada
    lista[pos] = elemento; 
    n++; // Incrementa o contador de elementos
}


/**
 * Remove um número da primeira posição da lista e movimenta os demais elementos para o início.
 * @return Valor do número que foi removido.
 */
int removerInicio() {

   // Verifica se a lista está vazia
    if (n == 0) {
        printf("Erro ao remover! Lista vazia.\n"); // Mensagem de erro se a lista está vazia
        exit(1); // Encerra o programa em caso de erro
    }

    int resp = lista[0]; // Guarda o número que será removido
    n--; // Decrementa o contador de elementos

    // Move os elementos restantes para o início do array
    for (int i = 0; i < n; i++) {
        lista[i] = lista[i + 1]; // Desloca cada número uma posição para a esquerda
    }

    return resp; // Retorna o número removido
}

/**
 * Remove um número da última posição da lista.
 * @return Valor do número que foi removido.
 */
int removerFim() {

   // Verifica se a lista está vazia
    if (n == 0) {
        printf("Erro ao remover! Lista vazia.\n"); // Mensagem de erro se a lista está vazia
        exit(1); // Encerra o programa em caso de erro
    }

    return lista[--n]; // Decrementa o contador e retorna o número da última posição
}

/**
 * Remove um número de uma posição específica da lista e movimenta os demais elementos para o início.
 * @param pos Posição de remoção.
 * @return Valor do número que foi removido.
 */
int remover(int pos) {
    // Verifica se a remoção é possível
    if (n == 0 || pos < 0 || pos >= n) {
        printf("Erro ao remover! Lista vazia ou posição inválida.\n"); // Mensagem de erro para lista vazia ou posição inválida
        exit(1); // Encerra o programa em caso de erro
    }

    int resp = lista[pos]; // Guarda o número que será removido
    n--; // Decrementa o contador de elementos

    // Move os elementos restantes para o início do array
    for (int i = pos; i < n; i++) {
        lista[i] = lista[i + 1]; // Desloca cada número uma posição para a esquerda
    }

    return resp; // Retorna o número removido
}

/**
 * Mostra todos os números armazenados na lista.
 */
void mostrar() {
    // Imprime cada número armazenado na lista
    printf("[ ");
    for (int i = 0; i < n; i++) {
        printf("%d ", lista[i]); // Exibe o valor do número
    }
    printf("]\n");
}

/**
 * Pesquisa um número na lista e retorna se ele foi encontrado.
 * @param elemento Valor do número a ser pesquisado.
 * @return true se o número foi encontrado; false caso contrário.
 */
bool pesquisar(int elemento) {
    bool retorno = false; // Inicializa a variável que indica se o número foi encontrado

    // Itera sobre a lista para procurar o número
    for (int i = 0; i < n; i++) {
        if (lista[i] == elemento) { // Verifica se o elemento atual é o número procurado
            retorno = true; // Define como true se o número for encontrado
            i = n; // Encerra a busca antecipadamente, pois o número foi encontrado
        }
    }
    
    return retorno; // Retorna true se o número foi encontrado, caso contrário retorna false
}

int main() {
    // Aqui você pode realizar testes chamando as funções de inserção e remoção

    return 0;
}